### PR TITLE
Return 503 for all interrupted queries. Refactor the query killing code.

### PR DESCRIPTION
**Pull request by jasperjiaguo**
1. Return 503 error code for killed/cancelled queries for both server and broker side interruption

2. Also change the default queriesKilled metric to not publish as this metric can have false alarms. The alternate approach is to alert on the return code.

3. Meanwhile moved the server initialization to server starter as it now does not depend on the pqw and pqr numbers.

4. When killing, include the instance id in the error message for observability

### CODE REVIEW

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java [hunk 1]
```diff
@@ -816,6 +816,9 @@ public static class Accounting {
 
     public static final String CONFIG_OF_GC_WAIT_TIME_MS = "accounting.gc.wait.time.ms";
     public static final int DEFAULT_CONFIG_OF_GC_WAIT_TIME_MS = 0;
+
+    public static final String CONFIG_OF_QUERY_KILLED_METRIC_ENABLED = "accounting.query.killed.metric.enabled";
+    public static final boolean DEFAULT_QUERY_KILLED_METRIC_ENABLED = false;
```
Comment (siddharthteotia) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:821]: Why is this change needed ?

Comment (siddharthteotia) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:821]: and why is this set to false ?

Comment (jasperjiaguo) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:821]: Because we want to emit this metric from our code by conditioning on the response code, as would have no false alert.

File: pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java [hunk 2]
```diff
@@ -545,6 +546,9 @@ public void start()
     InstanceDataManager instanceDataManager = _serverInstance.getInstanceDataManager();
     instanceDataManager.setSupplierOfIsServerReadyToServeQueries(() -> _isServerReadyToServeQueries);
     initSegmentFetcher(_serverConf);
+    // initialize the thread accountant for query killing
+    Tracing.ThreadAccountantOps
+        .initializeThreadAccountant(_serverConf.subset(CommonConstants.PINOT_QUERY_SCHEDULER_PREFIX), _instanceId);
```
Comment (walterddr) [inline pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java:551]: is this related to returning 205 instead of 200?

Comment (jasperjiaguo) [reply pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java:551]: Updated the pr description to include all the changes. PTAL.

**Comment by jasperjiaguo** [comment]
> if we are changing the status code, should we consider a general guideline of response code? at the moment error query result and success query result and partial query results are all returning 200 which could be misleading.
> 
> it is definitely bigger than the scope of this PR but would like to raise the concern before we determine 205 (Reset content) as the proper return code. according to document:
> 
> ```
> 205: The server successfully processed the request, asks that the requester reset its document view, and is not returning any content.
> ```
> 
> which doesn't seems like what's associated with query killing

@walterddr agreed, I think we should consider the error codes holistically sometime, it is indeed a bit misleading right now. This PR should not affect the future refactoring right? I don't think we depend on hard coded number here?

**Comment by walterddr** [comment]
> @walterddr agreed, I think we should consider the error codes holistically sometime, it is indeed a bit misleading right now. This PR should not affect the future refactoring right? I don't think we depend on hard coded number here?

hmm. i think we should agree upon the error code convension we wanted to use first. until then we should make this PR's query killing behavior return 200 with attach the exception on the JSON return payload just like the rest of the error query response. 
since the return code is used by the REST client it would be particularly difficult for backward-compatibility modifications going forward (say tomorrow we decided to return a 4xx instead of 205) some browser will consider non-2xx as exception instead of an acceptable return which will cause confusion IMO

**Comment by jasperjiaguo** [comment]
> > @walterddr agreed, I think we should consider the error codes holistically sometime, it is indeed a bit misleading right now. This PR should not affect the future refactoring right? I don't think we depend on hard coded number here?
> 
> hmm. i think we should agree upon the error code convension we wanted to use first. until then we should make this PR's query killing behavior return 200 with attach the exception on the JSON return payload just like the rest of the error query response. since the return code is used by the REST client it would be particularly difficult for backward-compatibility modifications going forward (say tomorrow we decided to return a 4xx instead of 205) some browser will consider non-2xx as exception instead of an acceptable return which will cause confusion IMO

Hey Rong, I completely understand your concern of backward-compatibility. We definitely want a consistency with the standard http error code for our query return codes going forward. However, there is still a possibility we do this in an incremental way for this query cancellation code, the reasons are

1.  Currently when a query is killed/cancelled on a broker it already returns 205 instead of 200, the following is from our log:
```
Got 1 errors for request: *****, 
Errors:
Error code: 205, Error message: org.apache.pinot.spi.exception.EarlyTerminationException:
```
IMO it's better for us to converge the code here and let server return the same for the interruption behavior

2. Currently we return all execution errors as 200 therefore some web browsers would take all execution failures as some kind of success, in this case, would changing from 200 -> 205 make it worse? Do you have deployments that depend on the hard-coded 200 for interruption errors?

This problem is currently impacting our deployment. PTAL cc @walterddr @siddharthteotia

**Comment by walterddr** [comment]
actually my concern was the other way around. sorry for miscommunication my concern.

2xx is consider [success](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes), this means 205 and 200 are both considered as success in HTTP standard. i was actually wondering if we should return an actual error code, for example
- 429 Too Many Requests
- 503 Service Unavailable
both seems to be more suitable for query killed ?

**Comment by jasperjiaguo** [comment]
Updated the ERROR code to 503 as @walterddr suggested
> The HyperText Transfer Protocol (HTTP) 503 Service Unavailable server error response code indicates that the server is not ready to handle the request.
> Common causes are a server that is down for maintenance or that is overloaded.

File: pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java [hunk 3]
```diff
@@ -54,7 +54,7 @@ public static void setMaxLinesOfStackTrace(int maxLinesOfStackTracePerFrame) {
   public static final int ACCESS_DENIED_ERROR_CODE = 180;
   public static final int TABLE_DOES_NOT_EXIST_ERROR_CODE = 190;
   public static final int QUERY_EXECUTION_ERROR_CODE = 200;
-  public static final int QUERY_CANCELLATION_ERROR_CODE = 205;
+  public static final int QUERY_CANCELLATION_ERROR_CODE = 503;
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java:57]: is it possible to add a unit test say similar to `CombineErrorOperatorsTest`?

**Review by walterddr — APPROVED** [review]
lgtm on the return code logic. will let Sidd take another look at the metrics pieces

File: pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java (hunk 3 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java:57]: As the error is returned from the runner thread not worker threads through the result block, having the error code test in OfflineClusterMemBasedServerQueryKilingTest  and OfflineClusterMemBasedBrokerQueryKilingTest and  would suffice

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java:57]: sounds good. it case should be good enough
